src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +138,129 @@ private void logTargetPath() {
         PathPlannerLogging.setLogTargetPoseCallback((pose) -> Logger.recordOutput("TargetPPPose", pose));
     }
 
+    private void addOdometryObservation(SwerveModulePosition[] swerveModulePositions, Rotation2d gyroHeading, double timestamp) {
+        final Twist2d odometryDifferenceTwist2d = calculateOdometryDifferenceTwist2d(swerveModulePositions, gyroHeading);
+        updateOdometryPositions(swerveModulePositions, gyroHeading);
+        updateRobotPosesFromOdometryDifferenceTwist2d(odometryDifferenceTwist2d, timestamp);
+    }
+
     private void updateFromVision() {
-        getViableVisionObservations().stream()
-                .sorted(Comparator.comparingDouble(PoseEstimator6328.VisionObservation::timestamp))
-                .forEach(poseEstimator6328::addVisionObservation);
+        updateAllCameras();
+        sortCamerasByLastTargetTimestamp();
+        addVisionObservations();
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private void updateAllCameras() {
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            aprilTagCamera.update();
         }
-        return viableVisionObservations;
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
+    private void sortCamerasByLastTargetTimestamp() {
+        VisionObservationSorter.quickSortByDoubleValue(aprilTagCameras, (aprilTagCamera) -> ((AprilTagCamera) aprilTagCamera).getLatestResultTimestampSeconds());
+    }
+
+    private void addVisionObservations() {
```

Comment:
1. add a javadoc
2. think of a better name. Maybe `updateVision` ?
3. Put this next to the odometry method. The way they are rn is annoying me.

Reply:
It uses the functions above it...

Reply:
Move one of the methods so that they'll be next to each other. I'm sure you can.